drivers/spi/spis_nrf5.c

```diff
@@ -0,0 +1,467 @@
+	rxd_amount = spi_regs->AMOUNTRX;
+#elif defined(CONFIG_SOC_SERIES_NRF52X)
+	txd_amount = spi_regs->RXD.AMOUNT;
+	rxd_amount = spi_regs->TXD.AMOUNT;
+#endif
+
+	SYS_LOG_DBG("bytes transferred: TX: %u, RX: %u [err %u (%s)]",
+		    txd_amount, rxd_amount,
+		    error, error == 0 ? "OK" : "ERR");
+
+	priv_data->error = error ? 1 : 0;
+
+	k_sem_give(&priv_data->device_sync_sem);
+}
+
+/**
+ * @brief SPI module interrupt handler.
+ * @param arg Pointer to the device structure for the driver instance
+ * @return None.
+ */
+static void spis_nrf5_isr(void *arg)
+{
+	struct device *dev = arg;
+	NRF_SPIS_Type *spi_regs = SPI_REGS(dev);
+	u32_t error;
+	u32_t tmp;
+
+	/* We get an interrupt for the following reasons:
+	 * 1. Semaphore ACQUIRED:
+	 *      Semaphore is assigned to the CPU again (always happens
+	 *      after END if the END_ACQUIRE SHORTS is set)
+	 * 2. End of Granted SPI transaction:
+	 *      Used to unblocked the caller, finishing the transaction
+	 */
+
+	/* NOTE:
+	 * Section 15.8.1 of nrf52 manual suggests reading back the register
+	 * to cause a 4-cycle delay to prevent the interrupt from
+	 * re-occurring
+	 */
+
+	if (spi_regs->EVENTS_END) {
+		spi_regs->EVENTS_END = 0;
+		/* force register flush (per spec) */
+		tmp = spi_regs->EVENTS_END;
+
+		/* Read and clear error flags. */
+		error = spi_regs->STATUS;
+		spi_regs->STATUS = error;
+
+		spis_nrf5_complete(dev, error);
+	}
+	if (spi_regs->EVENTS_ACQUIRED) {
+		spi_regs->EVENTS_ACQUIRED = 0;
+		/* force registesr flush (per spec) */
+		tmp = spi_regs->EVENTS_ACQUIRED;
+	}
+}
+
+static const struct spi_driver_api nrf5_spis_api = {
+	.transceive = spis_nrf5_transceive,
+	.configure = spis_nrf5_configure,
+	.slave_select = NULL,
+};
+
+#if defined(CONFIG_SOC_SERIES_NRF51X)
+static void spis_configure_psel(NRF_SPIS_Type *spi_regs,
+				const struct spis_nrf5_config *cfg)
+{
+	spi_regs->PSELMOSI = cfg->mosi_pin;
+	spi_regs->PSELMISO = cfg->miso_pin;
+	spi_regs->PSELSCK  = cfg->sck_pin;
+	if (cfg->csn_pin == NRF5_SPIS_CSN_DISABLED_CFG) {
+		spi_regs->PSELCSN = NRF5_SPIS_CSN_DISABLED;
+	} else {
+		spi_regs->PSELCSN = cfg->csn_pin;
+	}
+}
+#elif defined(CONFIG_SOC_SERIES_NRF52X)
+static void spis_configure_psel(NRF_SPIS_Type *spi_regs,
+				const struct spis_nrf5_config *cfg)
+{
+	spi_regs->PSEL.MOSI = cfg->mosi_pin;
+	spi_regs->PSEL.MISO = cfg->miso_pin;
+	spi_regs->PSEL.SCK = cfg->sck_pin;
+	if (cfg->csn_pin == NRF5_SPIS_CSN_DISABLED_CFG) {
+		spi_regs->PSEL.CSN = NRF5_SPIS_CSN_DISABLED;
+	} else {
+		spi_regs->PSEL.CSN = cfg->csn_pin;
+	}
+}
+#else
+#error "Unsupported NRF5 SoC"
+#endif
+
+static int spis_nrf5_init(struct device *dev)
+{
+	NRF_SPIS_Type *spi_regs = SPI_REGS(dev);
+	struct spis_nrf5_data *priv_data = DEV_DATA(dev);
+	const struct spis_nrf5_config *cfg = DEV_CFG(dev);
+	struct device *gpio_dev;
+	int ret;
+
+	SYS_LOG_DBG("SPI Slave driver init: %p", dev);
+
+	/* Enable constant latency for faster SPIS response */
+	NRF_POWER->TASKS_CONSTLAT = 1;
```

Comment:
This should be a new driver (power?), bind and enable; ideally Kconfig selectable.